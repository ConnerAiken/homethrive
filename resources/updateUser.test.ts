const updateUser = require('./../resources/updateUser');

describe('Update user actions PUT->/users/', () => {
    const user = {
        id: '123',
        firstName: 'Conner',
        lastName: 'Aiken',
        email: 'conner@fittedtech.com',
        dob: '06/02/1993',
    };
    const baseRequest = {
        resource: '/users',
        httpMethod: 'PUT',
    };

    test('We are able to update a user using the PUT endpoint', async () => {
        const requestPayload = {
            ...baseRequest,
            path: '/users',
            body: JSON.stringify(user),
        };

        const responsePayload = await updateUser(
            requestPayload,
            {
                update: jest.fn().mockReturnValue({
                    promise: () => {
                        return {
                            Attributes: user,
                        };
                    },
                }),
            },
            jest.fn().mockReturnValue({
                isValid: () => true,
            })
        );

        expect(responsePayload.statusCode).toBe(200);
        expect(responsePayload.body).toContain('Successfully updated');
        expect(JSON.parse(responsePayload.body)).toMatchSnapshot();
    });

    test('We send the proper server status code when there is an invalid body JSON', async () => {
        const requestPayload = {
            ...baseRequest,
            path: '/users',
            body: 'invalid json',
        };

        const responsePayload = await updateUser(
            requestPayload,
            {
                update: jest.fn().mockReturnValue({
                    promise: () => {
                        return {};
                    },
                }),
            },
            jest.fn().mockReturnValue({
                isValid: () => true,
            })
        );

        expect(responsePayload.statusCode).toBe(400);
        expect(responsePayload.body).toContain('Invalid JSON body');
    });

    test('We send the proper server status code when there is no relevant user', async () => {
        const requestPayload = {
            ...baseRequest,
            path: '/users',
            body: JSON.stringify(user),
        };

        const responsePayload = await updateUser(
            requestPayload,
            {
                update: jest.fn().mockReturnValue({
                    promise: () => {
                        return {};
                    },
                }),
            },
            jest.fn().mockReturnValue({
                isValid: () => true,
            })
        );

        expect(responsePayload.statusCode).toBe(404);
        expect(responsePayload.body).toContain('Could not find a user');
        expect(JSON.parse(responsePayload.body)).toMatchSnapshot();
    });

    test('We handle generic errors properly', async () => {
        const requestPayload = {
            ...baseRequest,
            path: '/users',
            body: JSON.stringify(user),
        };

        const responsePayload = await updateUser(
            requestPayload,
            {
                update: jest.fn().mockReturnValue({
                    promise: () => {
                        throw new Error('Simulated error');
                    },
                }),
            },
            jest.fn().mockReturnValue({
                isValid: () => true,
            })
        );

        expect(responsePayload.statusCode).toBe(500);
        expect(responsePayload.body).toContain('An error occured');
        expect(responsePayload.body).toMatchSnapshot();
    });
});
