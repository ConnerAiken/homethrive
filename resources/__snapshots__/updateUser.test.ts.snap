// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update user actions PUT->/users/ We are able to update a user using the PUT endpoint 1`] = `
{
  "error": false,
  "message": "Successfully updated the user",
  "payload": {
    "dob": "06/02/1993",
    "email": "conner@fittedtech.com",
    "firstName": "Conner",
    "id": "123",
    "lastName": "Aiken",
  },
}
`;

exports[`Update user actions PUT->/users/ We handle generic errors properly 1`] = `"{"message":"An error occured","payload":{},"error":{}}"`;

exports[`Update user actions PUT->/users/ We send the proper server status code when there is no relevant user 1`] = `
{
  "error": true,
  "message": "Could not find a user with the provided id",
  "payload": {},
}
`;
